Sign Up Logic with Clean Arch

Jimmybasha opened the pull request:
Added Api Call Using Dio
Added Dependency Injection for classes used using Injectable
Didn't Use Entity
Added Reusable Api Postdata For Further Extensions and uses
Modified Some failure Texts to accommodate the Sign Up errors
Added Bloc Builder for Error, Success Message
added another Bloc builder for the button Loading

Jimmybasha committed:
+SignUpLogic With Clean Architecture

john-safwat commented on lib/Core/ApiManager/ApiManager.dart:
```diff
@@ -1,7 +1,8 @@
 
 import '../Constants/Constants.dart';
-
+@injectable
```
api manager preferred to be singleton

john-safwat commented on lib/Core/Constants/Constants.dart:
```diff
@@ -1 +1,2 @@
-final String BASE_URL= "https://exam.elevateegy.com/api/v1/";
\ No newline at end of file
+final String BASE_URL= "https://exam.elevateegy.com/api/v1/";
+final String SIGNUP_URL="auth/signup";
```
it's preferred to follow the dart naming convention

john-safwat commented on lib/Core/utils/app_routes.dart:
```diff
@@ -1,8 +1,14 @@
   ForgetPasswordScreen.id: (context) => const ForgetPasswordScreen(),
   EmailVerificationScreen.id: (context) => const EmailVerificationScreen(),
+  SignUpPage.id: (context) => const SignUpPage(),
+  LoginScreen.id:(context) => const LoginScreen(),
```
use route generator for Clean Navigation

john-safwat commented on lib/Features/Auth/Login/presentation/view/widgets/DoNotHaveAccountWidget.dart:
```diff
@@ -18,14 +20,19 @@ class DoNotHaveAccountWidget extends StatelessWidget {
         ),
         GestureDetector(
-          onTap: () {},
+          onTap: () {
+            Navigator.pushNamed(context, SignUpPage.id);
+          },
+
           child: Text("Sign up",
```
this logic Is the responsibility of view model not the view

Jimmybasha replied:
it wasn't my part , it was my partner's part i'm responsible for the Sign Up

john-safwat commented on lib/Features/Auth/Login/presentation/view/widgets/DoNotHaveAccountWidget.dart:
```diff
@@ -18,14 +20,19 @@ class DoNotHaveAccountWidget extends StatelessWidget {
         ),
         GestureDetector(
-          onTap: () {},
+          onTap: () {
+            Navigator.pushNamed(context, SignUpPage.id);
+          },
+
           child: Text("Sign up",
```
don't use hard-coded Strings instead use localization or easy localization

Jimmybasha replied:
same here any Login Logic wasn't mine

john-safwat commented on lib/Features/Auth/Sign_Up/data/data_sources/Remote/SignUpDataSource.dart:
```diff
@@ -0,0 +1,14 @@
+import '../../../../../../Core/ApiManager/ApiResult.dart';
+import '../../models/UserSignUpModel.dart';
+
+abstract class SignUpDataSource{
```
use abstract interface class inter

john-safwat commented on lib/Features/Auth/Sign_Up/data/data_sources/Remote/SignUpDataSourceImpl.dart:
```diff
@@ -0,0 +1,44 @@
+import 'package:online_exam/Features/Auth/Sign_Up/data/data_sources/Remote/SignUpDataSource.dart';
+import 'package:online_exam/Features/Auth/Sign_Up/data/models/UserSignUpModel.dart';
+@Injectable(as: SignUpDataSource)
+class AuthRemoteDataSourceImpl extends SignUpDataSource{
```
remote data source is interface so you shouldn't extends from it

john-safwat commented on lib/Features/Auth/Sign_Up/presentation/View/widgets/AlreadyHaveAccountWidget.dart:
```diff
@@ -0,0 +1,34 @@
+    return  Row(
+      mainAxisAlignment: MainAxisAlignment.center,
+      children: [
+        Text("Already have an account?"),
```
don't use hard coded strings in your code

john-safwat commented on lib/Features/Auth/Sign_Up/presentation/View/widgets/AlreadyHaveAccountWidget.dart:
```diff
@@ -0,0 +1,34 @@
+            onPressed: (){
+
+            },
+            child: GestureDetector(
+              onTap: () {
+                Navigator.of(context).pushNamed(LoginScreen.id);
+              },
```
this logic is the responsibility of the view model

john-safwat commented on lib/Features/Auth/Sign_Up/presentation/View/widgets/AlreadyHaveAccountWidget.dart:
```diff
@@ -0,0 +1,34 @@
+              onTap: () {
+                Navigator.of(context).pushNamed(LoginScreen.id);
+              },
+              child: Text("Login",
```
don't use hard coded strings in your code

john-safwat commented on lib/Features/Auth/Sign_Up/presentation/View/widgets/SignUpScreenBody.dart:
```diff
@@ -134,40 +137,68 @@ class _SignUpScreenBodyState extends State<SignUpScreenBody> {
+                          height: 24,
+                          child: CircularProgressIndicator(color: Colors.white, strokeWidth: 2),
+                        )
+                            : const Text("Sign Up", style: TextStyle(color: Colors.white)),
```
don't use hard coded strings in your code

john-safwat commented on lib/Features/Auth/Sign_Up/presentation/View/widgets/SignUpScreenBody.dart:
```diff
@@ -134,40 +137,68 @@ class _SignUpScreenBodyState extends State<SignUpScreenBody> {
+                      WidgetsBinding.instance.addPostFrameCallback((_) {
+                        ScaffoldMessenger.of(context).showSnackBar(
+                          SnackBar(
+                            content: Text("Sign up Successfully"),
```
don't use hard coded strings in your code

john-safwat commented on lib/Features/Auth/Sign_Up/presentation/View_Model/cubits/SignUpViewModel.dart:
```diff
@@ -1,3 +1,55 @@
+      String rePassword,
+      )async{
+    emit(SignUpLoadingState());
+    var apiResult = await signUpUseCase.call(username, firstName, lastName, email, password, phone, rePassword);
+    print("APi Result in the signUpLoadingState apiResult : ${apiResult}");
+
+    switch(apiResult){
+      case SuccessApiResult():
+        print(" apiResult.Data in the SuccessApiResult :  ${apiResult.data?.user?.message??"Nothing appeared"}");
```
don't use prints in production code and instead use logger or debugPrint

Jimmybasha committed:
+Resolved the Print switched it into Logs
+Changed the Login Navigation Logic inside the signup view from the View to the viewModel
+Corrected the Hard Coded Strings
+Made The Navigator Function
+Added Navigator Key In the main to help in Navigation without using Context To Help in the ViewModel Navigation Logic
+ Added the Confirm password Validator
+Changed The extends to Implements (For Abstract Classes)
+Added ("I") Before Abstract Class to identify it as Interface
